[14.0][ADD] stock_location_orderpoint: location orderpoint replenish

mt-software-de opened the pull request:
Here i am adding a new module, which is allowing to define a orderpoint on a location. 
To replenish any product with the same criteria. 

This depends on:
- https://github.com/OCA/stock-logistics-warehouse/pull/1691
- https://github.com/OCA/wms/pull/616

mt-software-de committed:
[ADD] stock_location_orderpoint:

Declare orderpoint on a location allowing to replenish any product with the same criteria

mt-software-de commented:
@jbaudoux

rousseldenis commented:
@jbaudoux I will do a first migration of this to v16 in orderpoint repo.

jbaudoux commented on stock_location_orderpoint/models/stock_location_orderpoint.py:
```diff
@@ -0,0 +1,226 @@
+            ("procure_method", "=", "make_to_stock"),
+            ("location_id", "child_of", self.location_id.ids),
+            "!",
+            ("location_dest_id", "child_of", self.location_id.ids),
```
This is not safe to combine the locations of different orderpoints, you need to take each different location_id one by one and OR this subdomain for each location_id

jbaudoux commented on stock_location_orderpoint/models/stock_location_orderpoint.py:
```diff
@@ -0,0 +1,226 @@
+            "group_id": self.group_id,
+        }
+
+    def _get_replenishment_move_domain(self):
```
```suggestion
    def _get_waiting_move_domain(self):
```

jbaudoux commented on stock_location_orderpoint/models/stock_location_orderpoint.py:
```diff
@@ -0,0 +1,226 @@
+            ("location_dest_id", "child_of", self.location_id.ids),
+        ]
+
+    def _find_products_to_replenish_by_location(self):
```
A comment would help understanding what is being done

```suggestion
    def _find_**potential**_products_to_replenish_by_location(self):
        """ Return a dictionary of products per location that potentially require a replenishment based on the fact there are moves not reserved for those products. This reduces the list of products for which the quantity will be computed"""
```

jbaudoux commented on stock_location_orderpoint/models/stock_location_orderpoint.py:
```diff
@@ -0,0 +1,226 @@
+            for res in moves_grouped
+        }
+
+    def run_replenishment(self):
```
I would extract the computation of the (orderpoint, product, qty, proc_vals) to replenish in a separate method
and then keep here only the creation of the orderpoint and run

This is because looking at waiting moves and computing based on virtual available is one easy method when there is no min quantity defined that assumes a min=0
I would call this immediate replenishment.

We should allow extension with different methods that would returns different list based on different criteria.
Or maybe we need to add fields on the orderpoint to allow to configure the method for computing the min and max, and also a priority. This would allow different methods for a same location but resulting in moves with different priorities.

jbaudoux commented on stock_location_orderpoint/readme/DESCRIPTION.rst:
```diff
@@ -0,0 +1 @@
+Stock location orderpoint repelnish
```
```suggestion
Declare orderpoint on a location allowing to replenish any product with the same criteria.

The replenished quantity is limited to what is available at the source location to plan only reservable replenishment moves.
```

jbaudoux commented on stock_location_orderpoint/views/menu.xml:
```diff
@@ -0,0 +1,15 @@
+    <menuitem
+        id="menu_stock_location_orderpoint"
+        action="action_stock_location_orderpoint"
+        name="Stock Location Oderpoint"
```
```suggestion
        name="Stock Location Orderpoint"
```

jbaudoux commented on stock_location_orderpoint/models/stock_location_orderpoint.py:
```diff
@@ -0,0 +1,226 @@
+        self.ensure_one()
+        return {
+            "route_ids": self.route_id,
+            "date_planned": fields.Date.today(),
```
```suggestion
            "date_planned": fields.Datetime.now(),
```

jbaudoux commented on stock_location_orderpoint/readme/CONTRIBUTORS.rst:
```diff
@@ -0,0 +1 @@
+* Michael Tietz (MT Software) <mtietz@mt-software.de>
```
Can you add me also?

jbaudoux reviewed (changes requested):
Thanks for the implementation. This looks very good for a first version.

rousseldenis commented:
@mt-software-de Could you do separate commits for improvements (with a commit message like [IMP] stock_location_orderpoint: (...) )? That will allow me to forwardport them more easily ?

mt-software-de commented:
> @mt-software-de Could you do separate commits for improvements (with a commit message like [IMP] stock_location_orderpoint: (...) )? That will allow me to forwardport them more easily ?

I am not done with it yet, i have do some major improvements. 
After that i can do it. I think i'm done with in 2 hours.

mt-software-de commented on stock_location_orderpoint/models/stock_location_orderpoint.py:
```diff
@@ -0,0 +1,336 @@
+        default="auto",
+        required=True,
+    )
+    replenish_method = fields.Selection(
+        [("fill_up", "Fill up")],
+        default="fill_up",
+        required=True,
+        help="Defines how the qty to replenish is computed",
+    )
```
@jbaudoux or @rousseldenis do you have maybe a better name for this field and value.
It is used in _get_qty_to_replenish to get the right method to compute the qty to replenish

rousseldenis replied:
That seems good to me. What is bothering you ?

mt-software-de replied:
Is not bothering me, in German i would find an exact description for it, but in english if find sometimes hard.

mt-software-de commented:
> @mt-software-de Could you do separate commits for improvements (with a commit message like [IMP] stock_location_orderpoint: (...) )? That will allow me to forwardport them more easily ?

I am done with the major changes

jbaudoux commented on stock_location_orderpoint/models/stock_location_orderpoint.py:
```diff
@@ -0,0 +1,338 @@
+        domain='[("company_id", "=", company_id)]',
+    )
+    trigger = fields.Selection(
+        [("auto", "Auto"), ("manual", "Manual")],
```
I would maybe rename auto into scheduled (cron) and use auto/realtime based on new moves triggers.
I would generate the cron linked to this record and allow to define the periodicity on this model.

mt-software-de replied:
Yes cron would be better. i will do this.
But we can not implement one cron for each orderpoint.
This must be a computed cron because we can have multiple orderpoints for one location.

jbaudoux replied:
indeed, let's keep it like you did

jbaudoux commented on stock_location_orderpoint/models/stock_location_orderpoint.py:
```diff
@@ -0,0 +1,338 @@
+        self.ensure_one()
+        return {
+            "route_ids": self.route_id,
+            "date_planned": fields.Datetime.now(),
+            "date_deadline": False,
```
This needs to be refined.
For the implemented strategy, it should be the planned date of the oldest waiting move.
Then you can sort the replenishments by priority and have an accurate planned date

mt-software-de replied:
the orderpoints getting now sorted by "priority desc, sequence"
and date_planned is oldest date of the waiting moves for product and location

jbaudoux commented on stock_location_orderpoint/models/stock_location_orderpoint.py:
```diff
@@ -0,0 +1,338 @@
+            "date_planned": fields.Datetime.now(),
+            "date_deadline": False,
+            "warehouse_id": self.location_id.get_closest_warehouse(),
+            "group_id": self.group_id,
```
Nice to have: add a priority as defined on the orderpoint (field to add)

jbaudoux commented on stock_location_orderpoint/__manifest__.py:
```diff
@@ -0,0 +1,23 @@
+
+{
+    "name": "stock_location_orderpoint",
+    "author": "Michael Tietz (MT Software) <mtietz@mt-software.de>, "
```
```suggestion
    "author": "MT Software, "
```

jbaudoux reviewed (approved):
Thanks
I put some nice to have remarks.
This version looks now good enough, I approve it

jbaudoux commented on stock_location_orderpoint/models/stock_location_orderpoint.py:
```diff
@@ -0,0 +1,367 @@
+        domain='[("company_id", "=", company_id)]',
+    )
+    trigger = fields.Selection(
+        [("auto", "Auto/realtime"), ("manual", "Manual"), ("cron", "Scheduled")],
```
auto is not implemented yet, so maybe it should not be listed yet, or add a ROADMAP file with a note regarding this.
Then we can already merge this.

mt-software-de replied:
i will do the implementation next week, should we wait for it?

jbaudoux replied:
ok, great ! Add [WIP] in the title then

mt-software-de replied:
This is now done

jbaudoux commented on stock_location_orderpoint/models/stock_location_orderpoint.py:
```diff
@@ -0,0 +1,367 @@
+        default="auto",
+        required=True,
+        help="Auto/realtime orderpoints are triggered on new moves\n"
+        "Manual orderpoints are triggered via the orderpoint's view\n"
+        "Scheduled orderpoints are triggered vi scheduled actions per location",
```
```suggestion
        "Manual orderpoints are triggered via the orderpoints' view\n"
        "Scheduled orderpoints are triggered via scheduled actions per location",
```

jbaudoux commented on stock_location_orderpoint/models/stock_location_orderpoint.py:
```diff
@@ -0,0 +1,367 @@
+        default="fill_up",
+        required=True,
+        help="Defines how the qty to replenish gets computed\n"
+        "Fill up = The replenished quantity is limited to "
+        "what is available at the source location to plan only reservable replenishment moves",
```
```suggestion
        "Fill up = The replenishment will be triggered when a move is waiting availability and forecast quantity is negative at the location (i.e. min=0). The replenished quantity will bring back the forecast quantity to 0 (i.e. max=0) but will be limited to "
        "what is available at the source location to plan only reservable replenishment moves",
```

mt-software-de commented:
@rousseldenis should i create a PR forward port it to version 16.0?

jbaudoux commented:
> @rousseldenis should i create a PR forward port it to version 16.0?

Rewrite 2nd commit message to make first line short or merge it with first commit.
I'll review the new addition before doing the fwd port.

mt-software-de commented:
> > @rousseldenis should i create a PR forward port it to version 16.0?
> 
> Rewrite 2nd commit message to make first line short or merge it with first commit. I'll review the new addition before doing the fwd port.
Ok but then we can also merge the last commit into to first one

jbaudoux commented on stock_location_orderpoint/models/stock_location_orderpoint.py:
```diff
@@ -164,6 +164,8 @@ def _get_waiting_move_domain(self):
             ("state", "in", ["confirmed", "partially_available"]),
             ("move_orig_ids", "=", False),
             ("procure_method", "=", "make_to_stock"),
+            ("product_id.type", "=", "product"),
```
Why do you need this?

jbaudoux commented on stock_location_orderpoint/models/stock_location_orderpoint.py:
```diff
@@ -164,6 +164,8 @@ def _get_waiting_move_domain(self):
             ("move_orig_ids", "=", False),
             ("procure_method", "=", "make_to_stock"),
+            ("product_id.type", "=", "product"),
+            ("location_id.usage", "=", "internal"),
```
The location can also be a view

jbaudoux commented on stock_location_orderpoint/models/stock_move.py:
```diff
@@ -9,3 +11,66 @@ class StockMove(models.Model):
+
+    def _action_confirm(self, *args, **kwargs):
+        moves = super()._action_confirm(*args, **kwargs)
+        moves._prepare_location_orderpoint_replenishment(
```
Can you add a comment with the intent?
I understand this is to trigger replenishment for new moves waiting stock on the location to replenish.

jbaudoux commented on stock_location_orderpoint/models/stock_move.py:
```diff
@@ -9,3 +11,66 @@ class StockMove(models.Model):
+
+    def _action_done(self, *args, **kwargs):
+        moves = super()._action_done(*args, **kwargs)
+        moves._prepare_location_orderpoint_replenishment(
```
Can you add a comment with the intent?
Why not having the domain in a method like the other case?
This is to trigger when the stock increases on the location serving for the replenishment?
Shouldn't be on the move lines? Otherwise you need to include view location
I need to test, not clear to me

mt-software-de replied:
If a move is done which increases the stock on a source location of a configured orderpoints
it runs a replenishment for it

jbaudoux commented on stock_location_orderpoint/models/stock_move.py:
```diff
@@ -9,3 +11,66 @@ class StockMove(models.Model):
+            [
+                ("move_dest_ids", "=", False),
+                ("procure_method", "=", "make_to_stock"),
+                ("product_id.type", "=", "product"),
```
Why do you need this?

mt-software-de replied:
product_id.type ist not needed, yes
But the rest checks if there should be a replenishment happening if the move increases the stock

jbaudoux commented on stock_location_orderpoint/views/stock_location_orderpoint_views.xml:
```diff
@@ -0,0 +1,93 @@
+                        <field name="location_src_id" />
+                        <field name="route_id" />
+                        <field name="group_id" />
+                    </group>
```
Missing priority field

jbaudoux reviewed (approved):
Thanks for the last improvements & fixes

rousseldenis commented on stock_location_orderpoint/models/stock_location_orderpoint.py:
```diff
@@ -0,0 +1,416 @@
+        return {
+            "route_ids": self.route_id,
+            "date_deadline": False,
+            "warehouse_id": self.location_id.get_closest_warehouse(),
```
@mt-software-de Could you add a TODO here to remind me to remove that for v16 as warehouse_id field on locations is correctly implemented ?

jbaudoux replied:
No we cannot rely on odoo code for the computed warehouse because we have a virtual warehouse containing other warehouses:
WH/Stock (virtual warehouse for stock available to promise)
WH/Stock/MyWh1 (real warehouse1)
WH/Stock/MyWh2 (real warehouse2)
If the location to replenish is WH/Stock/MyWh1/Internal, then it must find the warehouse MyWh1 and not WH

rousseldenis replied:
That's what Odoo code in v16 do :smiley:

mt-software-de replied:
No i dont think we can remove it. 
Look at my answer Here. https://github.com/OCA/stock-logistics-warehouse/pull/1640#issuecomment-1451636719

rousseldenis reviewed (commented):
LGTM.

@mt-software-de Could you add the TODO and fix an error that has not been raised by github (see here under)?

Myabe that will fix some tests coverage

![image](https://user-images.githubusercontent.com/19529533/225850201-554b7114-55c4-4c80-b4a2-29747a467633.png)

mt-software-de commented:
> LGTM.
> 
> @mt-software-de Could you add the TODO and fix an error that has not been raised by github (see here under)?
> 
> Myabe that will fix some tests coverage
> 
> ![image](https://user-images.githubusercontent.com/19529533/225850201-554b7114-55c4-4c80-b4a2-29747a467633.png)

@rousseldenis where are you getting this error?
Because psycopg2 is by default a requirement of odoo

rousseldenis commented:
> > LGTM.
> > @mt-software-de Could you add the TODO and fix an error that has not been raised by github (see here under)?
> > Myabe that will fix some tests coverage
> > ![image](https://user-images.githubusercontent.com/19529533/225850201-554b7114-55c4-4c80-b4a2-29747a467633.png)
> 
> @rousseldenis where are you getting this error? Because psycopg2 is by default a requirement of odoo

Look in the tests here under

mt-software-de commented:
> > > LGTM.
> > > @mt-software-de Could you add the TODO and fix an error that has not been raised by github (see here under)?
> > > Myabe that will fix some tests coverage
> > > ![image](https://user-images.githubusercontent.com/19529533/225850201-554b7114-55c4-4c80-b4a2-29747a467633.png)
> > 
> > 
> > @rousseldenis where are you getting this error? Because psycopg2 is by default a requirement of odoo
> 
> Look in the tests here under

Sorry not where. :-)
On which system,  odoo version, python version, psycopg2 version?

rousseldenis commented:
> > > > LGTM.
> > > > @mt-software-de Could you add the TODO and fix an error that has not been raised by github (see here under)?
> > > > Myabe that will fix some tests coverage
> > > > ![image](https://user-images.githubusercontent.com/19529533/225850201-554b7114-55c4-4c80-b4a2-29747a467633.png)
> > > 
> > > 
> > > @rousseldenis where are you getting this error? Because psycopg2 is by default a requirement of odoo
> > 
> > 
> > Look in the tests here under
> 
> Sorry not where. :-) On which system, odoo version, python version, psycopg2 version?

I relaunched github tests, I think this is maybe a psycopg2 version problem

rousseldenis commented on stock_location_orderpoint/tests/test_location_orderpoint.py:
```diff
@@ -0,0 +1,296 @@
+# License AGPL-3.0 or later (https://www.gnu.org/licenses/agpl).
+from datetime import datetime
+
+from psycopg2.errors import UniqueViolation
```
@mt-software-de As psycopg2 2.7.7 is in the requirements of Odoo for python < 3.8, it is better to use:

`from psycopg2 import IntegrityError`

That writing is compliant with both versions

mt-software-de replied:
Yes this is better.

mt-software-de commented:
@jbaudoux i added you as author as well

jbaudoux commented:
@rousseldenis Can you update your review?

elvise commented:
@mt-software-de i’m getting an error when try to create a new rule, steps to reproduce 

1. Inventory 
2. Configuration
3. Stock location orderpoint 
4. Create
5. Error

Error:
Odoo Server Error

Traceback (most recent call last):
  File "/opt/odoo/odoo/addons/base/models/ir_http.py", line 237, in _dispatch
    result = request.dispatch()
  File "/opt/odoo/odoo/http.py", line 685, in dispatch
    result = self._call_function(**self.params)
  File "/opt/odoo/odoo/http.py", line 361, in _call_function
    return checked_call(self.db, *args, **kwargs)
  File "/opt/odoo/odoo/service/model.py", line 94, in wrapper
    return f(dbname, *args, **kwargs)
  File "/opt/odoo/odoo/http.py", line 349, in checked_call
    result = self.endpoint(*a, **kw)
  File "/opt/odoo/odoo/http.py", line 908, in __call__
    return self.method(*args, **kw)
  File "/opt/odoo/odoo/http.py", line 533, in response_wrap
    response = f(*args, **kw)
  File "/opt/odoo/addons/web/controllers/main.py", line 1372, in call_kw
    return self._call_kw(model, method, args, kwargs)
  File "/opt/odoo/addons/web/controllers/main.py", line 1364, in _call_kw
    return call_kw(request.env[model], method, args, kwargs)
  File "/opt/odoo/odoo/api.py", line 399, in call_kw
    result = _call_kw_multi(method, model, args, kwargs)
  File "/opt/odoo/odoo/api.py", line 386, in _call_kw_multi
    result = method(recs, *args, **kwargs)
  File "/opt/odoo/odoo/models.py", line 6295, in onchange
    snapshot1 = Snapshot(record, nametree)
  File "/opt/odoo/odoo/models.py", line 6086, in __init__
    self.fetch(name)
  File "/opt/odoo/odoo/models.py", line 6096, in fetch
    self[name] = record[name]
  File "/opt/odoo/odoo/models.py", line 5703, in __getitem__
    return self._fields[key].__get__(self, type(self))
  File "/opt/odoo/odoo/fields.py", line 2489, in __get__
    return super().__get__(records, owner)
  File "/opt/odoo/odoo/fields.py", line 969, in __get__
    self.recompute(record)
  File "/opt/odoo/odoo/fields.py", line 1155, in recompute
    self.compute_value(recs)
  File "/opt/odoo/odoo/fields.py", line 1177, in compute_value
    records._compute_field_value(self)
  File "/opt/odoo/odoo/models.py", line 4078, in _compute_field_value
    getattr(self, field.compute)()
  File "/mnt/data/odoo-addons-dir/stock_location_orderpoint/models/stock_location_orderpoint.py", line 130, in _compute_location_src_id
    orderpoint.location_src_id = orderpoint._get_location_src_id()
  File "/mnt/data/odoo-addons-dir/stock_location_orderpoint/models/stock_location_orderpoint.py", line 115, in _get_location_src_id
    "warehouse_id": self.location_id.get_closest_warehouse(),
  File "/mnt/data/odoo-addons-dir/stock_helper/models/stock_location.py", line 34, in get_closest_warehouse
    self.ensure_one()
  File "/opt/odoo/odoo/models.py", line 5028, in ensure_one
    raise ValueError("Expected singleton: %s" % self)
Exception

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "/opt/odoo/odoo/http.py", line 641, in _handle_exception
    return super(JsonRequest, self)._handle_exception(exception)
  File "/opt/odoo/odoo/http.py", line 317, in _handle_exception
    raise exception.with_traceback(None) from new_cause
ValueError: Expected singleton: stock.location()

mt-software-de commented:
> @mt-software-de i’m getting an error when try to create a new rule, steps to reproduce
> 
>     1. Inventory
> 
>     2. Configuration
> 
>     3. Stock location orderpoint
> 
>     4. Create
> 
>     5. Error
> 
> 
> Error: Odoo Server Error
> 
> Traceback (most recent call last): File "/opt/odoo/odoo/addons/base/models/ir_http.py", line 237, in _dispatch result = request.dispatch() File "/opt/odoo/odoo/http.py", line 685, in dispatch result = self._call_function(**self.params) File "/opt/odoo/odoo/http.py", line 361, in _call_function return checked_call(self.db, *args, **kwargs) File "/opt/odoo/odoo/service/model.py", line 94, in wrapper return f(dbname, *args, **kwargs) File "/opt/odoo/odoo/http.py", line 349, in checked_call result = self.endpoint(*a, **kw) File "/opt/odoo/odoo/http.py", line 908, in **call** return self.method(*args, **kw) File "/opt/odoo/odoo/http.py", line 533, in response_wrap response = f(*args, **kw) File "/opt/odoo/addons/web/controllers/main.py", line 1372, in call_kw return self._call_kw(model, method, args, kwargs) File "/opt/odoo/addons/web/controllers/main.py", line 1364, in _call_kw return call_kw(request.env[model], method, args, kwargs) File "/opt/odoo/odoo/api.py", line 399, in call_kw result = _call_kw_multi(method, model, args, kwargs) File "/opt/odoo/odoo/api.py", line 386, in _call_kw_multi result = method(recs, *args, **kwargs) File "/opt/odoo/odoo/models.py", line 6295, in onchange snapshot1 = Snapshot(record, nametree) File "/opt/odoo/odoo/models.py", line 6086, in **init** self.fetch(name) File "/opt/odoo/odoo/models.py", line 6096, in fetch self[name] = record[name] File "/opt/odoo/odoo/models.py", line 5703, in **getitem** return self._fields[key].**get**(self, type(self)) File "/opt/odoo/odoo/fields.py", line 2489, in **get** return super().**get**(records, owner) File "/opt/odoo/odoo/fields.py", line 969, in **get** self.recompute(record) File "/opt/odoo/odoo/fields.py", line 1155, in recompute self.compute_value(recs) File "/opt/odoo/odoo/fields.py", line 1177, in compute_value records._compute_field_value(self) File "/opt/odoo/odoo/models.py", line 4078, in _compute_field_value getattr(self, field.compute)() File "/mnt/data/odoo-addons-dir/stock_location_orderpoint/models/stock_location_orderpoint.py", line 130, in _compute_location_src_id orderpoint.location_src_id = orderpoint._get_location_src_id() File "/mnt/data/odoo-addons-dir/stock_location_orderpoint/models/stock_location_orderpoint.py", line 115, in _get_location_src_id "warehouse_id": self.location_id.get_closest_warehouse(), File "/mnt/data/odoo-addons-dir/stock_helper/models/stock_location.py", line 34, in get_closest_warehouse self.ensure_one() File "/opt/odoo/odoo/models.py", line 5028, in ensure_one raise ValueError("Expected singleton: %s" % self) Exception
> 
> The above exception was the direct cause of the following exception:
> 
> Traceback (most recent call last): File "/opt/odoo/odoo/http.py", line 641, in _handle_exception return super(JsonRequest, self)._handle_exception(exception) File "/opt/odoo/odoo/http.py", line 317, in _handle_exception raise exception.with_traceback(None) from new_cause ValueError: Expected singleton: stock.location()

Thank you very much. I fixed it

elvise commented:
@mt-software-de thanks for your fix!
Right now when click “Run Replenishment” getting this error:
Error:
Odoo Server Error

Traceback (most recent call last):
  File "/opt/odoo/odoo/addons/base/models/ir_http.py", line 237, in _dispatch
    result = request.dispatch()
  File "/opt/odoo/odoo/http.py", line 685, in dispatch
    result = self._call_function(**self.params)
  File "/opt/odoo/odoo/http.py", line 361, in _call_function
    return checked_call(self.db, *args, **kwargs)
  File "/opt/odoo/odoo/service/model.py", line 94, in wrapper
    return f(dbname, *args, **kwargs)
  File "/opt/odoo/odoo/http.py", line 349, in checked_call
    result = self.endpoint(*a, **kw)
  File "/opt/odoo/odoo/http.py", line 908, in __call__
    return self.method(*args, **kw)
  File "/opt/odoo/odoo/http.py", line 533, in response_wrap
    response = f(*args, **kw)
  File "/opt/odoo/addons/web/controllers/main.py", line 1376, in call_button
    action = self._call_kw(model, method, args, kwargs)
  File "/opt/odoo/addons/web/controllers/main.py", line 1364, in _call_kw
    return call_kw(request.env[model], method, args, kwargs)
  File "/opt/odoo/odoo/api.py", line 399, in call_kw
    result = _call_kw_multi(method, model, args, kwargs)
  File "/opt/odoo/odoo/api.py", line 386, in _call_kw_multi
    result = method(recs, *args, **kwargs)
  File "/mnt/data/odoo-addons-dir/stock_location_orderpoint/models/stock_location_orderpoint.py", line 336, in run_replenishment
    procurements = self._prepare_procurements(products)
  File "/mnt/data/odoo-addons-dir/stock_location_orderpoint/models/stock_location_orderpoint.py", line 332, in _prepare_procurements
    return self._sort_orderpoints().__prepare_procurements(moves_by_location)
  File "/mnt/data/odoo-addons-dir/stock_location_orderpoint/models/stock_location_orderpoint.py", line 310, in __prepare_procurements
    moves_by_location
  File "/mnt/data/odoo-addons-dir/stock_location_orderpoint/models/stock_location_orderpoint.py", line 292, in _get_qties_to_replenish
    qties_replenished_for_location[product],
  File "/mnt/data/odoo-addons-dir/stock_location_orderpoint/models/stock_location_orderpoint.py", line 239, in _get_qty_to_replenish
    product, qties_on_locations, qty_already_replenished
  File "/mnt/data/odoo-addons-dir/stock_location_orderpoint/models/stock_location_orderpoint.py", line 255, in _get_qty_to_replenish_fill_up
    qties_on_src = qties_on_locations[self.location_src_id][product]
Exception

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "/opt/odoo/odoo/http.py", line 641, in _handle_exception
    return super(JsonRequest, self)._handle_exception(exception)
  File "/opt/odoo/odoo/http.py", line 317, in _handle_exception
    raise exception.with_traceback(None) from new_cause
KeyError: stock.location()

mt-software-de commented:
> @mt-software-de thanks for your fix! Right now when click “Run Replenishment” getting this error: Error: Odoo Server Error
> 
> Traceback (most recent call last): File "/opt/odoo/odoo/addons/base/models/ir_http.py", line 237, in _dispatch result = request.dispatch() File "/opt/odoo/odoo/http.py", line 685, in dispatch result = self._call_function(**self.params) File "/opt/odoo/odoo/http.py", line 361, in _call_function return checked_call(self.db, *args, **kwargs) File "/opt/odoo/odoo/service/model.py", line 94, in wrapper return f(dbname, *args, **kwargs) File "/opt/odoo/odoo/http.py", line 349, in checked_call result = self.endpoint(*a, **kw) File "/opt/odoo/odoo/http.py", line 908, in **call** return self.method(*args, **kw) File "/opt/odoo/odoo/http.py", line 533, in response_wrap response = f(*args, **kw) File "/opt/odoo/addons/web/controllers/main.py", line 1376, in call_button action = self._call_kw(model, method, args, kwargs) File "/opt/odoo/addons/web/controllers/main.py", line 1364, in _call_kw return call_kw(request.env[model], method, args, kwargs) File "/opt/odoo/odoo/api.py", line 399, in call_kw result = _call_kw_multi(method, model, args, kwargs) File "/opt/odoo/odoo/api.py", line 386, in _call_kw_multi result = method(recs, *args, **kwargs) File "/mnt/data/odoo-addons-dir/stock_location_orderpoint/models/stock_location_orderpoint.py", line 336, in run_replenishment procurements = self._prepare_procurements(products) File "/mnt/data/odoo-addons-dir/stock_location_orderpoint/models/stock_location_orderpoint.py", line 332, in _prepare_procurements return self._sort_orderpoints().__prepare_procurements(moves_by_location) File "/mnt/data/odoo-addons-dir/stock_location_orderpoint/models/stock_location_orderpoint.py", line 310, in __prepare_procurements moves_by_location File "/mnt/data/odoo-addons-dir/stock_location_orderpoint/models/stock_location_orderpoint.py", line 292, in _get_qties_to_replenish qties_replenished_for_location[product], File "/mnt/data/odoo-addons-dir/stock_location_orderpoint/models/stock_location_orderpoint.py", line 239, in _get_qty_to_replenish product, qties_on_locations, qty_already_replenished File "/mnt/data/odoo-addons-dir/stock_location_orderpoint/models/stock_location_orderpoint.py", line 255, in _get_qty_to_replenish_fill_up qties_on_src = qties_on_locations[self.location_src_id][product] Exception
> 
> The above exception was the direct cause of the following exception:
> 
> Traceback (most recent call last): File "/opt/odoo/odoo/http.py", line 641, in _handle_exception return super(JsonRequest, self)._handle_exception(exception) File "/opt/odoo/odoo/http.py", line 317, in _handle_exception raise exception.with_traceback(None) from new_cause KeyError: stock.location()

Many thanks, are there not enough test from my site?
Ok i see the problems now my use case was only for internal replenishment,
so we always have a source location. 
I will add a fix to make it work.

mt-software-de commented:
Ok it is fixed, now.
But still something to do, with a buy route selected it will always add the qty to replenish to the purchase order line.

jbaudoux commented:
> Ok it is fixed, now. But still something to do, with a buy route selected it will always add the qty to replenish to the purchase order line.

@mt-software-de Don't squash new commits as it's being ported to v16

Can you log a note on the README this is for internal warehouse replenishment and not compatible with purchase buy route

mt-software-de commented:
> > Ok it is fixed, now. But still something to do, with a buy route selected it will always add the qty to replenish to the purchase order line.
> 
> @mt-software-de Don't squash new commits as it's being ported to v16
> 
> Can you log a note on the README this is for internal warehouse replenishment and not compatible with purchase buy route

since it was not yet merged and there was no notification that someone started the porting, i squashed the commits. 
If no one started the porting, i can do this as well. 
Should i create really an own commit for updating the README?

jbaudoux commented:
@rousseldenis Have you already started porting this one to v16? Can you answer @mt-software-de

mt-software-de commented:
> @rousseldenis Have you already started porting this one to v16? Can you answer @mt-software-de

Ping @rousseldenis, i would do the a new rebase with the info for the buy route and porting tomorrow if there is no answer from you, ok? :-) I would like to finish this :-)

mt-software-de committed:
[IMP] stock_move_auto_assign_auto_release: Fix tests assert_job_count

mt-software-de commented:
I started the forward porting. 
https://github.com/OCA/stock-logistics-warehouse/pull/1685

mt-software-de commented:
@jbaudoux and @rousseldenis could you update your review

rousseldenis commented:
@mt-software-de Could you add a magic button on location form level to access orderpoints like done on products level ?

rousseldenis commented on stock_location_orderpoint/models/stock_location_orderpoint.py:
```diff
@@ -0,0 +1,429 @@
+        ]
+        orderpoint_domains = []
+        for orderpoint in self:
+            _domain = [("origin", "=", orderpoint.name)]
```
I don't like this at all. You can not rely on a string name to search for records. For a reason X or Y, if the name has changed, this won't work.

You should use the location_orderpoint_id field instead.

So, you can add it before the loop `[(location_orderpoint_id', in, self.ids)]`

rousseldenis replied:
So, then, is there a need to add the location_id as we are strict enough with the orderpoint id ?

rousseldenis replied:
I've tested this and tests are passing (maybe some cases are not included ?):

```python
def _prepare_to_assign_replenishment_move_domain(self):
        """Returns a domain which selects moves created by a replenishment"""
        domain = [
            ("state", "in", ["confirmed", "partially_available"]),
            ("procure_method", "=", "make_to_stock"),
            ("location_orderpoint_id", "in", self.ids)
        ]
        return domain
```

mt-software-de replied:
This should be fine, i implemented the location_orderpoint_id field later in the process. 
Thats why there was the origin search. Thx

rousseldenis replied:
I don't agree on that as:

- Either you should add a unique constraint for the name too.
- 'This should be fine,' is not a jsutification. Could you elaborate a little bit more why you are searching on origin (maybe there is a good reason :smiley: )?

rousseldenis commented on stock_location_orderpoint/models/stock_location_orderpoint.py:
```diff
@@ -0,0 +1,429 @@
+    def _get_qties_to_replenish(self, moves_by_location):
+        products = self.env["product.product"]
+        for moves in moves_by_location.values():
+            products |= moves.product_id
```
As `|` is underperformant in ORM:

```suggestion
            product_ids.append(moves.product_id.id)
            qties_on_locations = self._compute_quantities_dict(
            (self.location_id | self.location_src_id), self.env["product.product"].browse(product_ids)
        )
```

rousseldenis commented on stock_location_orderpoint/models/stock_location_orderpoint.py:
```diff
@@ -0,0 +1,429 @@
+        for orderpoint, qties_to_replenish in qties_to_replenish_by_orderpoint.items():
+            proc_vals = orderpoint._prepare_procurement_values()
+            for product, qty in qties_to_replenish:
+                date_planned = min(
```
Could you move that part of code in separate method ?

I've tested that in stock.move with success:

```python
    @ormcache("moves", "product")
    def _get_location_orderpoint_replenishment_date(self, moves, product):
        return min(moves.filtered(
            lambda move: move.product_id == product).mapped("date"))
```

Then:

```suggestion
                date_planned = date_planned = move_obj._get_location_orderpoint_replenishment_date(
                    moves_by_location[orderpoint.location_id], product)
```

rousseldenis commented on stock_location_orderpoint/models/stock_move.py:
```diff
@@ -0,0 +1,79 @@
+                continue
+            location = getattr(move, location_field)
+            locations_products.setdefault(location, self.env["product.product"])
+            locations_products[location] |= move.product_id
```
Same as before, better to use ids than recordset with `|`

rousseldenis reviewed (commented):
Code review

mt-software-de committed:
[ADD] stock_location_orderpoint_source_relocate:

This module is a glue module between stock_location_orderpoint and stock_move_source_relocate
It will run an auto location orderpoint replenishment also after a move gets relocated

rousseldenis commented on stock_location_orderpoint/models/stock_location_orderpoint.py:
```diff
@@ -0,0 +1,423 @@
+        "If none is given, the moves generated by stock rules "
+        "will be grouped into one big picking.",
+    )
+    location_src_id = fields.Many2one(
```
@mt-software-de @jbaudoux If I understand correctly, this is always computed from the route. Could you add some documentation about how to configure orderpoints in USAGE.rst ? Thanks

Shouldn't the _get_location_src_id deserves a separate module for that purpose ?

mt-software-de replied:
Yes it is computed from the location_id and the route_id
I can move this to the stock_helper module

rousseldenis commented on stock_location_orderpoint/models/stock_location_orderpoint.py:
```diff
@@ -0,0 +1,423 @@
+    _sql_constraints = [
+        (
+            "location_route_unique",
+            "unique(location_id, route_id)",
```
@mt-software-de Please add company_id

mt-software-de replied:
Why? Shouldn't it be already company depended, because of check_company from the location_id field?

rousseldenis replied:
Indeed

rousseldenis commented on stock_location_orderpoint/models/stock_location_orderpoint.py:
```diff
@@ -0,0 +1,423 @@
+        "Name",
+        copy=False,
+        required=True,
+        readonly=True,
```
Remove this one as:

- One may want to change the name manually
- With readonly, the call to next_by_code is done twice : one at clicking on 'Create', one at Save. As the value is not transmitted through the interface. So, there is always a gap between two creations.

The other solution, maybe the best, is to add `force_save="1"` in views definition for the name field.

rousseldenis committed:
[IMP] stock_location_orderpoint: Add access to orderpoints from location form

rousseldenis commented:
@mt-software-de @jbaudoux I've added a shortcut on location form side:

https://github.com/mt-software-de/stock-logistics-warehouse/pull/2

rousseldenis commented:
@mt-software-de For the next time, changes in other modules should be done in separate PR's (which you can include with test-requirements.txt) because this impeaches using different ocabot commands (if major/minor in the other modules).

mt-software-de commented:
> @mt-software-de For the next time, changes in other modules should be done in separate PR's (which you can include with test-requirements.txt) because this impeaches using different ocabot commands (if major/minor in the other modules).

https://github.com/OCA/stock-logistics-warehouse/pull/1691

mt-software-de commented:
Ok. Now i think now we are ready. 
At first, these two PR, should be approved and merged. 
https://github.com/OCA/wms/pull/616
https://github.com/OCA/stock-logistics-warehouse/pull/1691
Then we can merge this one.

rousseldenis commented on stock_location_orderpoint/models/stock_move.py:
```diff
@@ -0,0 +1,87 @@
+
+        for location, products in locations_products.items():
+            for product in product_obj.browse(products):
+                self._enqueue_auto_replenishment(
```
@mt-software-de @jbaudoux 

After having tested the flow, I think we should not create queue jobs for stock locations that do not have an orderpoint.

Currently, every move validation is triggering a job. Useless IMHO

mt-software-de replied:
yes we do, i wanted to keep the queue.job preparation as simple as possible.
we can add after line 32, a check like
`if self.env["stock.location.orderpoint"]._get_orderpoints("auto", location, location_field)`

rousseldenis commented on stock_location_orderpoint/models/stock_move.py:
```diff
@@ -0,0 +1,87 @@
+    def _action_done(self, *args, **kwargs):
+        """
+        This triggers the replenishment for waiting moves
+        when the stock increases on a source location of an orderpoint
```
@mt-software-de In complement of my above comment, this does not do what it says.

All moves generate a job and not only for the ones that have an orderpoint !

mt-software-de commented on stock_location_orderpoint/models/stock_move.py:
```diff
@@ -0,0 +1,87 @@
+        )
+        return job
+
+    def _action_confirm(self, *args, **kwargs):
+        """This triggers the replenishment for new moves which are waiting for stock"""
+        moves = super()._action_confirm(*args, **kwargs)
+        moves._prepare_location_orderpoint_replenishment(
+            "location_id",
+            self.env["stock.location.orderpoint"]._get_waiting_move_domain(),
+        )
+        return moves
```
Maybe we should change this and use the action_assign method. 
And overwrite it via _regsiter_hook to be sure to execute/create the queue.job after the action_assign happens. 
Then we also don't need a glue module between stock_location_orderpoint and source_relocate. 
@jbaudoux and @rousseldenis what do you think, about this?

rousseldenis replied:
I'll look into. e.g.: try to do a manual incoming picking. After the first save and as the incoming pickings are automatically confirmed, a new job is created at each move creation.

mt-software-de replied:
@rousseldenis Could you have a look at. https://github.com/mt-software-de/stock-logistics-warehouse/pull/4

rousseldenis committed:
[IMP] stock_location_orderpoint: Add demo data to ease tests

rousseldenis committed:
[IMP] stock_location_orderpoint: Add CONFIGURE readme fragment

rousseldenis committed:
[IMP] stock_location_orderpoint: Improve loop and add a test

rousseldenis commented:
@mt-software-de Ok, I think we are almost at goal. Could you at least look at my PR https://github.com/mt-software-de/stock-logistics-warehouse/pull/5 ?

I think you could have avoided to squash all commits as they reflected the development.

mt-software-de commented:
> @mt-software-de Ok, I think we are almost at goal. Could you at least look at my PR [mt-software-de#5](https://github.com/mt-software-de/stock-logistics-warehouse/pull/5) ?
> 
> I think you could have avoided to squash all commits as they reflected the development.

I restored everything, cherry picked your last changes

rousseldenis commented:
@lmignon @jbaudoux This deserves a review. Thanks

mt-software-de commented:
> @lmignon @jbaudoux This deserves a review. Thanks

Please also review. 
https://github.com/OCA/wms/pull/616
https://github.com/OCA/stock-logistics-warehouse/pull/1691

lmignon commented on stock_location_orderpoint/models/stock_location_orderpoint.py:
```diff
@@ -0,0 +1,412 @@
+        self.ensure_one()
+        product.ensure_one()
+
+        qty_to_replinsh_func_name = f"_get_qty_to_replenish_{self.replenish_method}"
```
I'm not a fan of this kind of extension mechanism relaying on code introspection. It make code difficult to debug since it's not possible to easily find where a method is used (your ide  or the grep commande will never find where the `_get_qty_to_replenish_fill_up` method is used) Explicit is always better than implicit ant the code will also be easier to read....
```python
def _get_qty_to_replenish(
        self, product, qties_on_locations, qty_already_replenished=0
    ):
  if self.replenish_method == "fill_up":
    return self._get_qty_to_replenish_fill_up(product, qties_on_locations, qty_already_replenished)
  return 0
´´´

mt-software-de replied:
Thats fine for me,

lmignon commented on stock_location_orderpoint/models/stock_location_orderpoint.py:
```diff
@@ -0,0 +1,412 @@
+                    )
+                    qties_to_replenish[orderpoint].append((product, qty_to_replenish))
+                    qties_replenished_for_location[product] += qty_to_replenish
+        return qties_to_replenish
```
```suggestion
        from collections import defaultdict

        qties_replenished = defaultdict(lambda : defaultdict(lambda: 0))
        qties_to_replenish = defaultdict(list)
        for orderpoint in self:
            if orderpoint.location_id not in moves_by_location:
                continue

            for product in moves_by_location[orderpoint.location_id].product_id:
                qties_replenished_for_location = qties_replenished[orderpoint.location_id]
                qty_to_replenish = orderpoint._get_qty_to_replenish(
                    product,
                    qties_on_locations,
                    qties_replenished_for_location[product],
                )
                if (
                    float_compare(
                        qty_to_replenish, 0, precision_rounding=product.uom_id.rounding
                    )
                    > 0
                ):
                    qties_to_replenish[orderpoint].append((product, qty_to_replenish))
                    qties_replenished_for_location[product] += qty_to_replenish
        return qties_to_replenish
```

lmignon commented on stock_location_orderpoint/models/stock_move.py:
```diff
@@ -0,0 +1,103 @@
+            location_field == "location_id" and location_field or "location_src_id"
+        )
+        for location, products in locations_products.items():
+            orderpoints = self.env["stock.location.orderpoint"]._get_orderpoints(
```
This will generate 1 SQL qry by loop. The method should be called from outside the loop and called only once for all the locations....

mt-software-de replied:
If we calling it outside of the loop, then we have to do child_of check inside the loop, because the _get_orderpoints method is searching with parent_of.

lmignon replied:
@mt-software-de `child_of` check inside the loop is a simple python method `child_location.parent_path.startswith(parent_location.parent_path)`  It's far less expensive than an SQL query by loop...

lmignon reviewed (changes requested):
Thak you for this great addon! Some little comments but a blocking one (IMO) (Code review)

lmignon reviewed (approved):
LGTM (Code review)

rousseldenis commented:
@mt-software-de Could you remove your test-requirement commit as we can merge this ?

mt-software-de commented:
> @mt-software-de Could you remove your test-requirement commit as we can merge this ?

done

rousseldenis commented:
/ocabot merge patch

rousseldenis commented:
@mt-software-de Could you update v16 PR ? Thanks
